### PR TITLE
LEAN-2852: TTC-100 - Track the changes on the title - FE implementation

### PR DESCRIPTION
**Pull request by nghazzawi-atypon**

### CODE REVIEW

File: src/schema/nodes/title.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  tabindex: string
+}
+
+export interface TitleNode extends ManuscriptNode {
+  attrs: Attrs
+}
+
+export const title: NodeSpec = {
+  content: 'text*',
+  marks: '',
```
Comment (asouqi) [inline src/schema/nodes/title.ts:31]: can you please use marks in https://github.com/Atypon-OpenSource/manuscripts-title-editor/blob/master/src/schema.ts#L65C7-L65C12

File: src/schema/nodes/title.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+  attrs: {
+    tabindex: { default: '2' },
```
Comment (asouqi) [inline src/schema/nodes/title.ts:34]: for what we using this `tabindex`?

**Review by asouqi — CHANGES_REQUESTED** [review]

File: src/jats/importer/jats-body-dom-parser.ts [hunk 3]
```diff
@@ -702,6 +702,33 @@ const nodes: NodeRule[] = [
     node: 'caption_title',
     context: 'figcaption/',
   },
+  {
```
Comment (mbartenev-atypon) [inline src/jats/importer/jats-body-dom-parser.ts:705]: Shouldn't we remove the old way we did that around src/jats/importer/parse-jats-article.ts:93?

File: src/schema/nodes/title.ts (hunk 1 above)
Comment (nghazzawi-atypon) [reply src/schema/nodes/title.ts]: done

File: src/schema/nodes/title.ts (hunk 2 above)
Comment (nghazzawi-atypon) [reply src/schema/nodes/title.ts]: removed

File: src/transformer/decode.ts [hunk 4]
```diff
@@ -150,6 +152,9 @@ const getAffiliations = (modelMap: Map<string, Model>) =>
 const getContributors = (modelMap: Map<string, Model>) =>
   getModelsByType<Affiliation>(modelMap, ObjectTypes.Contributor)
 
+const getTitle = (modelMap: Map<string, Model>) =>
+  getModelsByType<Title>(modelMap, ObjectTypes.Title)[0]
```
Comment (asouqi) [inline src/transformer/decode.ts:156]: you could replace this with `findTitle` or use `getModelsByType(model).find...`, as assuming there will always be a MPTitle could break decoding if it's missing

File: src/transformer/project-bundle.ts [hunk 5]
```diff
@@ -110,3 +111,15 @@ export const findManuscriptById = (
...
+const isTitle = hasObjectType<Title>(ObjectTypes.Title)
+
+export const findTitle = (modelMap: Map<string, Model>): Title => {
+  for (const model of modelMap.values()) {
+    if (isTitle(model)) {
+      return model
+    }
+  }
+
+  throw new Error('No Title found')
```
Comment (asouqi) [inline src/transformer/project-bundle.ts:124]: I think in this case we should have a default title Model. CC @mbartenev-atypon

**Review by asouqi — CHANGES_REQUESTED** [review]
We need to move the title node to the top of the manuscript node

File: src/transformer/decode.ts (hunk 4 above)
Comment (nghazzawi-atypon) [reply src/transformer/decode.ts:156]: I'm doing a check on line line 825.. but I will replace it with findTitle
image

File: src/transformer/project-bundle.ts (hunk 5 above)
Comment (nghazzawi-atypon) [reply src/transformer/project-bundle.ts]: done

File: src/transformer/decode.ts (hunk 4 above)
Comment (nghazzawi-atypon) [reply src/transformer/decode.ts]: I'm doing a check on line 825

File: src/transformer/builders.ts [hunk 6]
```diff
@@ -94,10 +94,9 @@ export const buildProject = (owner: string): Build<Project> => ({
   title: '',
 })
 
-export const buildManuscript = (title = ''): Build<Manuscript> => ({
+export const buildManuscript = (): Build<Manuscript> => ({
   _id: generateID(ObjectTypes.Manuscript),
   objectType: ObjectTypes.Manuscript,
-  title,
 })
```
Comment (asouqi) [inline src/transformer/builders.ts:101]: can you please add this title builder
```
export const buildTitles = (articleTitle?: string): Build<Titles> => ({
  _id: generateID(ObjectTypes.ElementsOrder),
  objectType: ObjectTypes.Titles,
  title: articleTitle || '',
})
```

File: src/transformer/decode.ts [hunk 7]
```diff
@@ -182,6 +184,16 @@ export class Decoder {
...
   private creators: NodeCreatorMap = {
+    [ObjectTypes.Titles]: (data) => {
+      const model = data as Titles
+
+      return schema.nodes.titles.create({
+        id: model._id,
+        title: model.title,
+        subtitle: model.subtitle,
+        runningTitle: model.runningTitle,
+      }) as TitlesNode
```
Comment (asouqi) [inline src/transformer/decode.ts:195]: it need to be like like this:
```
      return this.parseContents(model.title, 'div', undefined, {
        topNode: schema.nodes.titles.create({
          id: model._id,
          subtitle: model.subtitle,
          runningTitle: model.runningTitle,
        }),
      })
```

File: src/transformer/decode.ts [hunk 8]
```diff
@@ -872,6 +884,19 @@ export class Decoder {
...
+      ObjectTypes.Titles
+    )[0]
+
+    if (titlesModel) {
+      const titlesNode = this.decode(titlesModel) as TitlesNode
+      return titlesNode
+    } else {
+      return null
+    }
+  }
```
Comment (asouqi) [inline src/transformer/decode.ts:899]: simplify to be:
```
  private createTitlesNode() {
    const titlesModel =
      getModelsByType<Titles>(this.modelMap, ObjectTypes.Titles)?.at(0) ||
      (buildTitles() as Titles)

    return this.decode(titlesModel) as TitlesNode
  }
```

File: src/transformer/decode.ts [hunk 9]
```diff
@@ -930,10 +955,15 @@ export class Decoder {
...
   public createArticleNode = (manuscriptID?: string): ManuscriptNode => {
+    const titlesNode = this.createTitlesNode()
     const rootSectionNodes = this.createRootSectionNodes()
     const metaSectionNode = this.createMetaSectionNode()
     const contents: ManuscriptNode[] = [...rootSectionNodes, metaSectionNode]
 
+    if (titlesNode) {
+      contents.push(titlesNode)
+    }
+
```
Comment (asouqi) [inline src/transformer/decode.ts:966]: remove this

File: src/transformer/decode.ts [hunk 10]
```diff
@@ -930,10 +955,15 @@ export class Decoder {
     this.modelMap.get(id) as T | undefined
 
   public createArticleNode = (manuscriptID?: string): ManuscriptNode => {
+    const titlesNode = this.createTitlesNode()
     const rootSectionNodes = this.createRootSectionNodes()
     const metaSectionNode = this.createMetaSectionNode()
     const contents: ManuscriptNode[] = [...rootSectionNodes, metaSectionNode]
```
Comment (asouqi) [inline src/transformer/decode.ts:961]: add titleNode to content:
```
 const contents: ManuscriptNode[] = [
      titlesNode,
      ...rootSectionNodes,
      metaSectionNode,
    ]
```

File: src/schema/nodes/titles.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  attrs: Attrs
+}
+
+export const titles: NodeSpec = {
+  content: 'text*',
+  marks: 'italic smallcaps subscript superscript',
+
+  attrs: {
+    id: { default: '' },
+    title: { default: '' },
```
Comment (asouqi) [inline src/schema/nodes/titles.ts:38]: I don't think we need this, as we agreed to keep it in the node content as text

File: src/schema/nodes/titles.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  runningTitle: string
+}
+
+export interface TitlesNode extends ManuscriptNode {
+  attrs: Attrs
+}
+
+export const titles: NodeSpec = {
+  content: 'text*',
+  marks: 'italic smallcaps subscript superscript',
```
Comment (asouqi) [inline src/schema/nodes/titles.ts:34]: you will need to add this marks: `tracked_insert tracked_delete`

File: src/schema/nodes/titles.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  attrs: {
+    id: { default: '' },
+    title: { default: '' },
+    subtitle: { default: '' },
+    runningTitle: { default: '' },
+    dataTracked: { default: null },
+  },
+  group: 'block element',
+  parseDOM: [{ tag: 'div' }],
+  toDOM: () => ['div', 0],
```
Comment (asouqi) [inline src/schema/nodes/titles.ts:45]: can you please add class to the node so you can style it. 
```
toDOM: () => ['div' , { class: 'kind elementIndex' }, 0]
```

**Review by asouqi — APPROVED** [review]

**Pull request closed without merging**